File: scripts/python/configurator.py
import argparse, json

def main():
    parser = argparse.ArgumentParser(
                prog='DCS Olympus configurator',
                description='This software allows to edit the DCS Olympus configuration file',
                epilog='')

    parser.add_argument('-a', '--address')
    parser.add_argument('-c', '--clientPort')
    parser.add_argument('-b', '--backendPort')
    parser.add_argument('-p', '--password')
    parser.add_argument('-bp', '--bluePassword')
    parser.add_argument('-rp', '--redPassword')

    args = parser.parse_args()

    with open("olympus.json", "r") as fp:
        config = json.load(fp)

    if (args.address is not None):
        config["server"]["address"] = args.address
        print(f"Address set to {args.address}")
    else:
        print("No address provided, skipping...")

    if args.backendPort is not None:
        if args.backendPort.isdecimal():
            config["server"]["port"] = int(args.backendPort)
            print(f"Backend port set to {args.backendPort}")
        else:
            print(f"Invalid backend port provided {args.backendPort}")
    else:
        print("No backend port provided, skipping...")

    if (args.password is not None):
        config["authentication"]["gameMasterPassword"] = args.password
        print(f"Game Master password set to {args.password}")
    else:
        print("No Game Master password provided, skipping...")

    if (args.bluePassword is not None):
        config["authentication"]["blueCommanderPassword"] = args.bluePassword
        print(f"Blue Commander password set to {args.bluePassword}")
    else:
        print("No Blue Commander password provided, skipping...")

    if (args.redPassword is not None):
        config["authentication"]["redCommanderPassword"] = args.redPassword
        print(f"Red Commander password set to {args.redPassword}")
    else:
        print("No Red Commander password provided, skipping...")

    if args.clientPort is not None:
        if args.clientPort.isdecimal():
            config["client"]["port"] = int(args.clientPort)
            print(f"Client port set to {args.clientPort}")
        else:
            print(f"Invalid client port provided {args.clientPort}")
    else:
        print("No client port provided, skipping...")

    with open("olympus.json", "w") as fp:
        json.dump(config, fp, indent = 4)

if __name__ == "__main__":
    main()

File: configurator/configurator/configurator.cpp
// configurator.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <fstream>
#include <nlohmann/json.hpp>
#include "argparse/argparse.hpp"
#include "sha256.h"
#include "windows.h"
#include <chrono>
#include <thread>


using json = nlohmann::json;
using namespace std;

struct MyArgs : public argparse::Args {
    string& address = kwarg("a", "Address");
    int& clientPort = kwarg("c", "Client port");
    int& backendPort = kwarg("b", "Backend port");
    string& gameMasterPassword = kwarg("p", "Game master password");
    string& blueCommanderPassword = kwarg("bp", "Blue commander password");
    string& redCommanderPassword = kwarg("rp", "Red commander password");
};

void SetStdinEcho(bool enable = true)
{
    HANDLE hStdin = GetStdHandle(STD_INPUT_HANDLE);
    DWORD mode;
    GetConsoleMode(hStdin, &mode);

    if (!enable)
        mode &= ~ENABLE_ECHO_INPUT;
    else
        mode |= ENABLE_ECHO_INPUT;

    SetConsoleMode(hStdin, mode);
}

int main(int argc, char* argv[]) 
{
    cout << "DCS Olympus configurator v1.0.0" << endl << endl << endl;

    SHA256 sha256;

    std::ifstream f("olympus.json");
    json data = json::parse(f);

    string address = data["server"]["address"];
    int clientPort = data["client"]["port"];
    int backendPort = data["server"]["port"];
    string gameMasterPassword = data["authentication"]["gameMasterPassword"];
    string blueCommanderPassword = data["authentication"]["blueCommanderPassword"];
    string redCommanderPassword = data["authentication"]["redCommanderPassword"];

    if (argc > 1) {
        auto args = argparse::parse<MyArgs>(argc, argv);
        args.print();      // prints all variables

        address = args.address;
        clientPort = args.clientPort;
        backendPort = args.backendPort;
        gameMasterPassword = sha256(args.gameMasterPassword);
        blueCommanderPassword = sha256(args.blueCommanderPassword);
        redCommanderPassword = sha256(args.redCommanderPassword);

    } else {
        cout << "No arguments provided, running in interactive mode" << endl;

        SetStdinEcho(true);

        string newValue;

        cout << "Insert a new address or press Enter to keep current value: " << address << endl << ">";
        getline(cin, newValue);
        if (newValue != "")
            address = newValue;

        while (true) {
            cout << "Insert a new client port or press Enter to keep current value: " << clientPort << endl << ">";
            getline(cin, newValue);
            if (newValue != "") {
                try {
                    clientPort = stoi(newValue);
                    break;
                }
                catch (...) {
                    cout << "Insert a valid integer number" << endl;
                }
            }
            else {
                break;
            }
        }

        while (true) {
            cout << "Insert a new backend port or press Enter to keep current value: " << backendPort << endl << ">";
            getline(cin, newValue);
            if (newValue != "") {
                try {
                    backendPort = stoi(newValue);
                    break;
                }
                catch (...) {
                    cout << "Insert a valid integer number" << endl;
                }
            }
            else {
                break;
            }
        };

        SetStdinEcho(false);

        while (true) {
            cout << "Insert a new Game Master password" << endl << ">";
            getline(cin, newValue);
            cout << endl;
            if (newValue != "") {
                gameMasterPassword = sha256(newValue);
                break;
            }
            else {
                cout << "Value can't be empty" << endl;
            }
        }

        while (true) {
            cout << "Insert a new Blue Commander password" << endl << ">";
            getline(cin, newValue);
            cout << endl;
            if (newValue != "") {
                blueCommanderPassword = sha256(newValue);
                break;
            }
            else {
                cout << "Value can't be empty" << endl;
            }
        }

        while (true) {
            cout << "Insert a new Red Commander password" << endl << ">";
            getline(cin, newValue);
            cout << endl;
            if (newValue != "") {
                redCommanderPassword = sha256(newValue);
                break;
            }
            else {
                cout << "Value can't be empty" << endl;
            }
        }
    }

    data["server"]["address"] = address;
    data["server"]["port"] = backendPort;
    data["authentication"]["gameMasterPassword"] = gameMasterPassword;
    data["authentication"]["blueCommanderPassword"] = blueCommanderPassword;
    data["authentication"]["redCommanderPassword"] = redCommanderPassword;
    data["client"]["port"] = clientPort;

    std::ofstream o("olympus.json");
    o << std::setw(4) << data << std::endl;

    cout << endl;
    cout << "Configuration updated successfully, bye!" << endl;

    this_thread::sleep_for(chrono::milliseconds(2000));
}

